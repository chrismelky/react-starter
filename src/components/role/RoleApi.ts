import axios from 'axios';

const resourceUrl = 'api/roles';

export const getRoles = async (opts?: any): Promise<any> => {
  const res = await axios.get(resourceUrl, { params: opts });
  return res.data;
};
