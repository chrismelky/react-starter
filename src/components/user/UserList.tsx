import React, { useState } from 'react';
import { DataTable } from 'primereact/datatable';
import { Button } from 'primereact/button';
import { Column } from 'primereact/column';
import { Dialog } from 'primereact/dialog';
import { User } from './User';
import { UserUpdate } from './UserUpdate';
import { Card } from 'primereact/card';
import { useQuery } from 'react-query';
import { getUsers } from './UserApi';

export default function UserList() {
  const [per_page] = useState<number>(5);

  const [page] = useState<number>(1);

  const { isLoading, data, isError, refetch } = useQuery(
    ['users', page, per_page],
    () => getUsers({ page, per_page }),
    {
      keepPreviousData: true,
      retry: false,
    },
  );

  const [showCreateOrUpdate, setShowCreateOrUpdate] = useState(false);

  const [user, setUser] = useState<User | null>();

  const createOrEdit = (user?: User) => {
    const data = user || {
      ...new User(),
    };
    setUser(data);
    setShowCreateOrUpdate(true);
  };

  const closeDialog = (result: boolean) => {
    setUser(null);
    setShowCreateOrUpdate(false);
    if (result) {
    }
  };

  const actions = (rowData: User) => {
    return (
      <React.Fragment>
        <Button
          icon="pi pi-pencil"
          className="p-button-rounded p-button-success mr-2"
          onClick={() => createOrEdit(rowData)}
        />
      </React.Fragment>
    );
  };

  return (
    <div className="flex flex-column gap-3">
      <Card></Card>
      <Card style={{ width: '100%' }}>
        {isError ? (
          <span>
            Error loading Data please referesh{' '}
            <Button onClick={() => refetch()} label="Reload"></Button>
          </span>
        ) : (
          <DataTable
            totalRecords={data?.total}
            paginator={true}
            rows={per_page}
            value={data?.data}
            loading={isLoading}
            style={{ width: '100%' }}>
            <Column field="first_name" header="First Name" />
            <Column field="last_name" header="Last Name" />
            <Column header="Email" field="email" />
            <Column body={actions} header="" />
          </DataTable>
        )}

        <Dialog
          onHide={() => closeDialog(false)}
          visible={showCreateOrUpdate}
          style={{ width: '450px' }}
          header="Create or Update User"
          modal>
          <UserUpdate user={user} onClose={closeDialog} />
        </Dialog>
      </Card>
    </div>
  );
}
