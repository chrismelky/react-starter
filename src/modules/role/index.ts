import RoleList from './role-list';

export * from './role';
export * from './role-api';
export * from './role-list';
export * from './role-update';

export default RoleList;
