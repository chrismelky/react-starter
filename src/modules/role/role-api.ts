import axios from 'axios';
import { useMutation, useQuery } from 'react-query';
import { Role } from '.';
import { createRequestParams, IApiParams } from '../../utils/utils';

const resourceUrl = 'api/roles';

/**
 * fetch Roles by page, and optional filters
 * @param IApiParams
 * @returns UseQueryResult
 */
export const useFetchRoles = ({
  queryParams,
  onSuccess,
  onError,
}: IApiParams) => {
  return useQuery<any>(
    ['fetchRoles', queryParams],
    () =>
      axios.get(`${resourceUrl}`, { params: createRequestParams(queryParams) }),
    {
      onSuccess,
      onError,
    },
  );
};

/**
 * create new Role or updated existing Role
 * Role to be created or updated is passed to mutate function param
 * @param IApiParams
 * @returns UseMutationResult
 */
export const useCreateOrUpdateRole = ({ onSuccess, onError }: IApiParams) => {
  return useMutation<Role, any, Role>(
    'createRole',
    (role: Role) => (role.id ? updateRole(role) : createRole(role)),
    {
      onSuccess,
      onError,
    },
  );
};

/**
 * Delete Role
 * @param IApiParams
 * @returns  UseMutationResult
 */
export const useDeleteRole = ({ onSuccess, onError }: IApiParams) => {
  return useMutation(
    'deteleRole',
    (id: number) => axios.delete(`${resourceUrl}/${id}`),
    {
      onSuccess,
      onError,
    },
  );
};

const createRole = (role: Role): any => {
  return axios.post(`${resourceUrl}`, role);
};

const updateRole = (role: Role): any => {
  return axios.patch(`${resourceUrl}/${role.id}`, role);
};
