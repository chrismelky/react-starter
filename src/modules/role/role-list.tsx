import React, { useEffect, useRef, useState } from 'react';
import { DataTable, DataTableFilterMeta } from 'primereact/datatable';
import { Button } from 'primereact/button';
import { Column } from 'primereact/column';
import { Toast } from 'primereact/toast';
import { ConfirmDialog, confirmDialog } from 'primereact/confirmdialog';
import { Card } from 'primereact/card';
import { ErrorFetching, AppTableAction } from '../shared';
import ErrorBoundary from '../../utils/error-boundary';
import {
  Role,
  RoleUpdate,
  useDeleteRole,
  useFetchRoles,
  roleDefaultValue,
} from '.';
import {
  IQueryParams,
  PAGE_SIZE_OPTIONS,
  stringDefaultFilter,
} from '../../utils/utils';

export default function RoleList() {
  //Config column filters
  const initialOptionFilters: DataTableFilterMeta = {
    name: stringDefaultFilter,
  };

  const pageSizeOptions = PAGE_SIZE_OPTIONS;

  const [showCreateOrUpdate, setShowCreateOrUpdate] = useState(false);

  //Selected Role to be updated or new Role
  const [role, setRole] = useState<Role>();

  /** a separeate filter state to avoid data loading on initilization, table columns that can be filtered and they options (i.e not necessary to fetch data)*/
  const [optionalFilters, setOptionalFilters] = useState<DataTableFilterMeta>();

  const toast = useRef<Toast>(null);

  /** pagination option, optional filters ,
   *  useFetchRoles function observe this state and reload data when any property value is changed
   */
  const [queryParams, setQueryParams] = useState<IQueryParams>({
    first: 0,
    pageSize: 10,
    page: 0,
    optionalFilters: undefined,
    sortField: undefined,
    sortOrder: undefined,
  });

  const {
    isLoading,
    isError,
    error: errorFetching,
    refetch,
    data: roles,
  } = useFetchRoles({
    queryParams,
  });

  const initFilters = () => {
    setOptionalFilters(initialOptionFilters);
  };

  // When filter cleared init filter state and update queryParam filter property
  const clearFilters = () => {
    initFilters();
    setQueryParams({
      ...queryParams,
      optionalFilters,
    });
  };

  useEffect(() => {
    initFilters();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const deteleMutation = useDeleteRole({
    onSuccess: () => {
      refetch();
      toast.current?.show({
        severity: 'success',
        summary: 'Deleted successfully',
        detail: 'Role deleted',
        life: 3000,
      });
    },
  });

  const createOrEdit = (roleData?: Role) => {
    const data = roleData || roleDefaultValue;
    setRole(data);
    setShowCreateOrUpdate(true);
  };

  const closeDialog = (result: boolean) => {
    setShowCreateOrUpdate(false);
    setRole(undefined);
    if (result) {
      refetch();
      toast.current?.show({
        severity: 'success',
        summary: 'Success Message',
        detail: 'Role created successfully',
        life: 3000,
      });
    }
  };

  const pageChanges = (event: any) => {
    setQueryParams({
      ...queryParams,
      page: event.page,
      first: event.first,
      pageSize: event.rows,
    });
  };

  const onSort = (event) => {
    setQueryParams({
      ...queryParams,
      sortField: event.sortField,
      sortOrder: event.sortOrder,
    });
  };

  /**When filter change update queryParam filter property */
  const onFilter = (event: any) => {
    if (Object.keys(event.filters).length) {
      event['first'] = 0;
      setQueryParams({
        ...queryParams,
        optionalFilters: { ...event.filters },
      });
    }
  };

  const confirmDelete = (role: Role) => {
    confirmDialog({
      header: 'Confirm Detele Role',
      message: 'Are you sure you want to delete this Role',
      accept: () => deteleMutation.mutate(role.id!),
    });
  };

  const header = () => {
    return (
      <div className="flex flex-row justify-content-between align-items-center">
        <span className="text-lg">Role</span>

        <div className="flex flex-row justify-content-start align-items-center gap-1">
          <Button
            type="button"
            icon="pi pi-filter-slash"
            label="Clear filter"
            className="p-button-text p-button-plain"
            onClick={clearFilters}
          />
          <Button
            icon="pi pi-plus"
            className="p-button-raised"
            onClick={() => createOrEdit()}
            label="Create Role"></Button>
        </div>
      </div>
    );
  };

  const actions = (rowData: Role) => {
    return (
      <AppTableAction
        rowData={rowData}
        edit={createOrEdit}
        confirmDelete={confirmDelete}
      />
    );
  };

  return (
    <div className="flex flex-column gap-2">
      <Card style={{ width: '100%' }}>
        {isError ? (
          <ErrorFetching error={errorFetching} refetch={refetch} />
        ) : (
          <DataTable
            header={header}
            size="small"
            lazy
            paginator
            first={queryParams.first}
            rows={queryParams.pageSize}
            totalRecords={roles?.total}
            onPage={pageChanges}
            filters={optionalFilters}
            rowsPerPageOptions={pageSizeOptions}
            filterDisplay="menu"
            dataKey="id"
            onFilter={onFilter}
            value={roles?.data}
            loading={isLoading}
            onSort={onSort}
            sortField={queryParams.sortField}
            sortOrder={queryParams.sortOrder}
            style={{ width: '100%' }}>
            <Column
              filter
              sortable
              filterPlaceholder="Search by Name"
              field="name"
              header="Name"
            />
            <Column className="actions" body={actions} />
          </DataTable>
        )}
        {showCreateOrUpdate && (
          <ErrorBoundary>
            <RoleUpdate
              show={showCreateOrUpdate}
              role={role}
              onClose={closeDialog}
            />
          </ErrorBoundary>
        )}
      </Card>
      <Toast ref={toast} />
      <ConfirmDialog />
    </div>
  );
}
