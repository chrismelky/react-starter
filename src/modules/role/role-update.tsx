import React from 'react';
import { Controller, useForm } from 'react-hook-form';
import { Button } from 'primereact/button';
import { InputText } from 'primereact/inputtext';
import { classNames } from 'primereact/utils';
import { Dialog } from 'primereact/dialog';
import { Message } from 'primereact/message';

import { Role, useCreateOrUpdateRole } from '.';

export const RoleUpdate = ({ role, onClose, show }: any) => {
  const {
    control,
    formState: { errors },
    handleSubmit,
  } = useForm({
    defaultValues: role,
  });

  const {
    mutate: createOrUpdate,
    isError,
    error,
  } = useCreateOrUpdateRole({
    onSuccess: () => {
      onClose(true);
    },
  });

  const onSubmit = async (data: Role) => {
    createOrUpdate(data);
  };

  const getFormErrorMessage = (name: string) => {
    return (
      errors[name] && <small className="p-error">{errors[name].message}</small>
    );
  };

  return (
    <Dialog
      onHide={() => onClose(false)}
      visible={show}
      style={{ width: '450px' }}
      header="Create or Update Role"
      modal>
      <div>
        <form onSubmit={handleSubmit(onSubmit)} className="p-fluid">
          <div className="flex flex-column gap-2 mt-2">
            {isError && (
              <Message
                severity="error"
                content={
                  typeof error.message === 'object'
                    ? error.message?.map((e: string) => <li key={e}>{e}</li>)
                    : error.message
                }></Message>
            )}

            <div className="field">
              <span className="p-float-label">
                <Controller
                  name="name"
                  control={control}
                  rules={{
                    required: 'Name is required',
                  }}
                  render={({ field, fieldState }) => (
                    <InputText id={field.name} {...field} />
                  )}
                />
                <label
                  htmlFor="name"
                  className={classNames({ 'p-error': errors.name })}>
                  Name
                </label>
              </span>
              {getFormErrorMessage('name')}
            </div>
          </div>
          <Button type="submit" label="Submit" className="mt-2" />
        </form>
      </div>
    </Dialog>
  );
};
