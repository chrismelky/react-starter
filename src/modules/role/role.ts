export class Role {
  public id?: number | null;
  public name?: string;
}

export const roleDefaultValue: Readonly<Role> = {
  id: null,
  name: '',
};
