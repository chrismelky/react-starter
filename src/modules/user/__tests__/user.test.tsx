import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import UserList, { useCreateOrUpdateUser, useFetchUsers } from '..';
import { useFetchRoles } from '../../role';
import renderer from 'react-test-renderer';

const mockUseFetchUsers = useFetchUsers as jest.Mock<any>;
const mockUseCreateOrUpdateUser = useCreateOrUpdateUser as jest.Mock<any>;
const mockUseFetchRoles = useFetchRoles as jest.Mock<any>;

jest.mock('../user-api');
jest.mock('../../role/role-api');

describe('UserList', () => {
  beforeEach(() => {
    mockUseFetchUsers.mockImplementation(() => ({
      isLoading: true,
      data: {
        data: [
          {
            firstName: 'TestFirstName',
          },
        ],
      },
      isError: false,
    }));

    mockUseFetchRoles.mockImplementation(() => ({
      isLoading: true,
      data: {
        data: [
          {
            id: 1,
            name: 'roleOne',
          },
        ],
      },
      isError: false,
    }));

    mockUseCreateOrUpdateUser.mockImplementation(() => ({
      mutate: jest.fn(() => {}),
      isError: false,
    }));
  });

  test('Render List of users', () => {
    render(<UserList />);
    const linkElement = screen.getByText(/TestFirstName/i);
    expect(linkElement).toBeInTheDocument();
  });

  test('Display Create/Update Popup', () => {
    render(<UserList />);
    fireEvent.click(screen.getByTestId('btn-create'));
    expect(screen.getByTestId(/create-update-dialog/i)).toBeInTheDocument();
    expect(screen.getByTestId(/roles/i)).toBeInTheDocument();
  });

  test('Render Correct', () => {
    const tree = renderer.create(<UserList />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
