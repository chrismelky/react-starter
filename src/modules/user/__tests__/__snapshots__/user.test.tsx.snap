// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList Render Correct 1`] = `
<div
  className="flex flex-column gap-2"
>
  <div
    className="p-card p-component"
    id={null}
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="p-card-body"
    >
      <div
        className="p-card-content"
      >
        <div
          className="p-datatable p-component p-datatable-responsive-stack p-datatable-gridlines p-datatable-sm"
          data-scrollselectors=".p-datatable-wrapper"
          id={null}
          style={
            Object {
              "width": "100%",
            }
          }
        >
          <div
            className="p-datatable-loading-overlay p-component-overlay"
          >
            <i
              className="p-datatable-loading-icon pi-spin pi pi-spinner"
            />
          </div>
          <div
            className="p-datatable-header"
          >
            <div
              className="flex flex-row justify-content-between align-items-center"
            >
              <span
                className="text-lg"
              >
                Users
              </span>
              <div
                className="flex flex-row justify-content-start align-items-center gap-1"
              >
                <button
                  aria-label="Clear filter"
                  className="p-button p-component p-button-text p-button-plain"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    className="p-button-icon p-c p-button-icon-left pi pi-filter-slash"
                  />
                  <span
                    className="p-button-label p-c"
                  >
                    Clear filter
                  </span>
                </button>
                <button
                  aria-label="Create User"
                  className="p-button p-component p-button-raised"
                  data-testid="btn-create"
                  disabled={false}
                  onClick={[Function]}
                >
                  <span
                    className="p-button-icon p-c p-button-icon-left pi pi-plus"
                  />
                  <span
                    className="p-button-label p-c"
                  >
                    Create User
                  </span>
                </button>
              </div>
            </div>
          </div>
          <div
            className="p-datatable-wrapper"
            style={
              Object {
                "maxHeight": null,
              }
            }
          >
            <table
              className="p-datatable-table"
              role="table"
              style={null}
            >
              <thead
                className="p-datatable-thead"
              >
                <tr
                  role="row"
                >
                  <th
                    aria-sort="none"
                    className="p-sortable-column"
                    colSpan={null}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDragStart={[Function]}
                    onDrop={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    role="columnheader"
                    rowSpan={null}
                    style={Object {}}
                    tabIndex={0}
                  >
                    <div
                      className="p-column-header-content"
                    >
                      <span
                        className="p-column-title"
                      >
                        First Name
                      </span>
                      <span
                        className="p-sortable-column-icon pi pi-fw pi-sort-alt"
                      />
                      <div
                        className="p-column-filter p-fluid p-column-filter-menu"
                      >
                        <button
                          aria-expanded={false}
                          aria-haspopup={true}
                          className="p-column-filter-menu-button p-link"
                          onClick={[Function]}
                          onKeyDown={[Function]}
                          type="button"
                        >
                          <span
                            className="pi pi-filter-icon pi-filter"
                          />
                        </button>
                      </div>
                    </div>
                  </th>
                  <th
                    aria-sort="none"
                    className="p-sortable-column"
                    colSpan={null}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDragStart={[Function]}
                    onDrop={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    role="columnheader"
                    rowSpan={null}
                    style={Object {}}
                    tabIndex={0}
                  >
                    <div
                      className="p-column-header-content"
                    >
                      <span
                        className="p-column-title"
                      >
                        Last Name
                      </span>
                      <span
                        className="p-sortable-column-icon pi pi-fw pi-sort-alt"
                      />
                      <div
                        className="p-column-filter p-fluid p-column-filter-menu"
                      >
                        <button
                          aria-expanded={false}
                          aria-haspopup={true}
                          className="p-column-filter-menu-button p-link"
                          onClick={[Function]}
                          onKeyDown={[Function]}
                          type="button"
                        >
                          <span
                            className="pi pi-filter-icon pi-filter"
                          />
                        </button>
                      </div>
                    </div>
                  </th>
                  <th
                    aria-sort={null}
                    className=""
                    colSpan={null}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDragStart={[Function]}
                    onDrop={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    role="columnheader"
                    rowSpan={null}
                    style={Object {}}
                    tabIndex={null}
                  >
                    <div
                      className="p-column-header-content"
                    >
                      <span
                        className="p-column-title"
                      >
                        Email
                      </span>
                    </div>
                  </th>
                  <th
                    aria-sort={null}
                    className="actions"
                    colSpan={null}
                    onClick={[Function]}
                    onDragLeave={[Function]}
                    onDragOver={[Function]}
                    onDragStart={[Function]}
                    onDrop={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    role="columnheader"
                    rowSpan={null}
                    style={Object {}}
                    tabIndex={null}
                  >
                    <div
                      className="p-column-header-content"
                    >
                      <span
                        className="p-column-title"
                      />
                    </div>
                  </th>
                </tr>
              </thead>
              <tbody
                className="p-datatable-tbody"
              >
                <tr
                  className=""
                  onClick={[Function]}
                  onContextMenu={[Function]}
                  onDoubleClick={[Function]}
                  onDragEnd={[Function]}
                  onDragLeave={[Function]}
                  onDragOver={[Function]}
                  onDragStart={[Function]}
                  onDrop={[Function]}
                  onKeyDown={[Function]}
                  onMouseDown={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  role="row"
                  tabIndex={null}
                >
                  <td
                    className=""
                    onBlur={[Function]}
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    onMouseUp={[Function]}
                    role="cell"
                    rowSpan={null}
                    style={Object {}}
                    tabIndex={null}
                  >
                    <span
                      className="p-column-title"
                    >
                      First Name
                    </span>
                    TestFirstName
                  </td>
                  <td
                    className=""
                    onBlur={[Function]}
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    onMouseUp={[Function]}
                    role="cell"
                    rowSpan={null}
                    style={Object {}}
                    tabIndex={null}
                  >
                    <span
                      className="p-column-title"
                    >
                      Last Name
                    </span>
                  </td>
                  <td
                    className=""
                    onBlur={[Function]}
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    onMouseUp={[Function]}
                    role="cell"
                    rowSpan={null}
                    style={Object {}}
                    tabIndex={null}
                  >
                    <span
                      className="p-column-title"
                    >
                      Email
                    </span>
                  </td>
                  <td
                    className="actions"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    onMouseDown={[Function]}
                    onMouseUp={[Function]}
                    role="cell"
                    rowSpan={null}
                    style={Object {}}
                    tabIndex={null}
                  >
                    <span
                      className="p-column-title"
                    />
                    <button
                      className="p-button p-component p-button-plain p-button-rounded p-button-text p-button-icon-only p-button-icon-only"
                      disabled={false}
                      onClick={[Function]}
                    >
                      <span
                        className="p-button-icon p-c pi pi-pencil"
                      />
                      <span
                        className="p-button-label p-c"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "&nbsp;",
                          }
                        }
                      />
                    </button>
                    <button
                      className="p-button p-component p-button-plain p-button-rounded p-button-text p-button-icon-only p-button-icon-only"
                      disabled={false}
                      onClick={[Function]}
                    >
                      <span
                        className="p-button-icon p-c pi pi-trash"
                      />
                      <span
                        className="p-button-label p-c"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "&nbsp;",
                          }
                        }
                      />
                    </button>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
          <div
            className="p-paginator p-component p-paginator-bottom"
            style={null}
          >
            <button
              aria-label="First Page"
              className="p-button p-component p-paginator-first p-paginator-element p-link p-disabled p-button-icon-only p-disabled"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <span
                className="p-button-icon p-c p-paginator-icon pi pi-angle-double-left"
              />
              <span
                className="p-button-label p-c"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&nbsp;",
                  }
                }
              />
            </button>
            <button
              aria-label="Previous Page"
              className="p-button p-component p-paginator-prev p-paginator-element p-link p-disabled p-button-icon-only p-disabled"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <span
                className="p-button-icon p-c p-paginator-icon pi pi-angle-left"
              />
              <span
                className="p-button-label p-c"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&nbsp;",
                  }
                }
              />
            </button>
            <span
              className="p-paginator-pages"
            />
            <button
              aria-label="Next Page"
              className="p-button p-component p-paginator-next p-paginator-element p-link p-disabled p-button-icon-only p-disabled"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <span
                className="p-button-icon p-c p-paginator-icon pi pi-angle-right"
              />
              <span
                className="p-button-label p-c"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&nbsp;",
                  }
                }
              />
            </button>
            <button
              aria-label="Last Page"
              className="p-button p-component p-paginator-last p-paginator-element p-link p-disabled p-button-icon-only p-disabled"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              <span
                className="p-button-icon p-c p-paginator-icon pi pi-angle-double-right"
              />
              <span
                className="p-button-label p-c"
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&nbsp;",
                  }
                }
              />
            </button>
            <div
              className="p-dropdown p-component p-inputwrapper p-disabled p-inputwrapper-filled"
              id={null}
              onClick={[Function]}
              onContextMenu={null}
              onMouseDown={null}
              style={null}
            >
              <div
                className="p-hidden-accessible"
              >
                <input
                  aria-haspopup="listbox"
                  aria-label={null}
                  aria-labelledby={null}
                  disabled={true}
                  id={null}
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  readOnly={true}
                  tabIndex={null}
                  type="text"
                />
              </div>
              <div
                className="p-hidden-accessible p-dropdown-hidden-select"
              >
                <select
                  aria-hidden="true"
                  name={null}
                  required={false}
                  tabIndex={-1}
                >
                  <option
                    value=""
                  />
                  <option
                    value={10}
                  >
                    10
                  </option>
                </select>
              </div>
              <span
                className="p-dropdown-label p-inputtext"
              >
                10
              </span>
              <div
                aria-expanded={false}
                aria-haspopup="listbox"
                className="p-dropdown-trigger"
                role="button"
              >
                <span
                  className="p-dropdown-trigger-icon p-clickable pi pi-chevron-down"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
